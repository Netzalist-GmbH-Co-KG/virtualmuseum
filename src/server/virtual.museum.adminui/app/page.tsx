import { Card, CardContent, CardDescription, CardHeader, CardTitle } from "@/components/ui/card"
import { DoorOpen, Clock, Film, FileImage } from "lucide-react"
import Link from "next/link"

export default function Dashboard() {
  const stats = [
    {
      title: "Rooms",
      value: "5",
      description: "Museum spaces & inventory",
      icon: DoorOpen,
      href: "/rooms",
    },
    {
      title: "Time Series",
      value: "8",
      description: "Historical sequences",
      icon: Clock,
      href: "/time-series",
    },
    {
      title: "Presentations",
      value: "15",
      description: "Multimedia shows",
      icon: Film,
      href: "/presentations",
    },
    {
      title: "Media Files",
      value: "42",
      description: "Images, videos & audio",
      icon: FileImage,
      href: "/media",
    },
  ]

  return (
    <div className="flex-1 space-y-4 p-8 pt-6">
      <div className="flex items-center justify-between space-y-2">
        <h2 className="text-3xl font-bold tracking-tight">Dashboard</h2>
      </div>
      <div className="grid gap-4 md:grid-cols-2 lg:grid-cols-2">
        {stats.map((stat) => (
          <Link href={stat.href} key={stat.title}>
            <Card className="hover:bg-accent/50 transition-colors cursor-pointer">
              <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
                <CardTitle className="text-sm font-medium">{stat.title}</CardTitle>
                <stat.icon className="h-4 w-4 text-muted-foreground" />
              </CardHeader>
              <CardContent>
                <div className="text-2xl font-bold">{stat.value}</div>
                <p className="text-xs text-muted-foreground mt-1">{stat.description}</p>
              </CardContent>
            </Card>
          </Link>
        ))}
      </div>
      <div className="grid gap-4 grid-cols-1 md:grid-cols-2 lg:grid-cols-3">
        <Card className="col-span-full md:col-span-2">
          <CardHeader>
            <CardTitle>TimeGlideVR Museum Management</CardTitle>
            <CardDescription>Administrative interface for managing VR museum content</CardDescription>
          </CardHeader>
          <CardContent>
            <p>
              Welcome to the TimeGlideVR Museum Management System. This interface allows museum staff to edit time
              series data and multimedia presentations for the VR application.
            </p>
            <p className="mt-4">
              Use the navigation menu to access different sections of the application. You can manage rooms and their
              inventory, time series, presentations, and media files.
            </p>
          </CardContent>
        </Card>
        <Card>
          <CardHeader>
            <CardTitle>Quick Tips</CardTitle>
          </CardHeader>
          <CardContent>
            <ul className="list-disc pl-5 space-y-2">
              <li>Create rooms first, then add inventory items to them</li>
              <li>Create time series before adding geo events</li>
              <li>Upload media files before creating presentations</li>
              <li>Remember that Slot 0 is for audio only</li>
              <li>Slot 1 is for 360Â° content</li>
              <li>Slots 2+ are for flat displays</li>
            </ul>
          </CardContent>
        </Card>
      </div>
    </div>
  )
}
