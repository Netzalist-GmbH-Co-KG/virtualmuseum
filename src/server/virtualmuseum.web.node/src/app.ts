import express, { Express, NextFunction, Request, Response } from 'express';
import cors from 'cors';
import swaggerUi from 'swagger-ui-express';
import { specs } from './config/swagger';
import { authMiddleware } from './middleware/auth.middleware';
import { DatabaseService } from './services/database.service';
import { ConfigService } from './services/config.service';
import { TenantRouter } from './routes/tenant.routes';
import { TimeSeriesRouter } from './routes/timeseries.routes';
import { MultiMediaPresentationRouter } from './routes/multimediapresentations.routes';

export class App {
    private app: Express;
    private configService: ConfigService;
    private dbService: DatabaseService;

    constructor(dbService?: DatabaseService) {
        this.configService = ConfigService.getInstance();
        this.dbService = dbService || new DatabaseService(this.configService.get('dbPath'));
        this.app = express();
        this.initializeMiddlewares();
    }

    async initialize(): Promise<void> {
        await this.dbService.initialize();
        this.initializeRoutes();
        this.setupErrorHandling();
    }

    private initializeMiddlewares(): void {
        this.app.use(cors());
        this.app.use(express.json());
        this.app.use('/swagger', swaggerUi.serve, swaggerUi.setup(specs));
        this.app.use('/api', authMiddleware);
    }

    private initializeRoutes(): void {
        // Health check endpoint (no auth required)
        this.app.get('/health', (req, res) => {
            res.json({ status: 'healthy' });
        });

        // API routes
        const tenantRouter = new TenantRouter(this.dbService);
        const timeSeriesRouter = new TimeSeriesRouter(this.dbService);
        const multiMediaPresentationRouter = new MultiMediaPresentationRouter(this.dbService);

        this.app.use('/api/tenants', tenantRouter.getRouter());
        this.app.use('/api/timeseries', timeSeriesRouter.getRouter());
        this.app.use('/api/multimediapresentations', multiMediaPresentationRouter.getRouter());
    }

    private setupErrorHandling(): void {
        // Handle 404
        this.app.use((req: Request, res: Response) => {
            res.status(404).json({ error: 'Not Found' });
        });

        // Handle all other errors
        this.app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
            console.error('Error:', err);
            res.status(500).json({ 
                error: 'Internal Server Error',
                message: err.message,
                stack: process.env.NODE_ENV === 'development' ? err.stack : undefined
            });
        });
    }

    public async start(): Promise<void> {
        try {
            await this.initialize();
            const port = parseInt(this.configService.get('PORT') || '3000', 10);
            const host = this.configService.get('HOST') || '0.0.0.0';

            this.app.listen(port, host, () => {
                console.log(`Server is running on http://localhost:${port}`);
                console.log(`Swagger documentation available at http://localhost:${port}/swagger`);
            });
        } catch (error) {
            console.error('Failed to start the application:', error);
            process.exit(1);
        }
    }

    public getApp(): Express {
        return this.app;
    }

    async close(): Promise<void> {
        await this.dbService.close();
    }
}
