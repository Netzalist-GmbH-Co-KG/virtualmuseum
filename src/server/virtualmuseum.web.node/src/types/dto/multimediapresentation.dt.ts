import { MultimediaPresentation, PresentationItem, MediaFile } from '../';


export interface PresentationItemWithMediaFile extends PresentationItem {
    MediaFile: MediaFile
}

export interface MultimediaPresentationWithPresentationItems extends MultimediaPresentation {
    PresentationItems: PresentationItemWithMediaFile[]
}
