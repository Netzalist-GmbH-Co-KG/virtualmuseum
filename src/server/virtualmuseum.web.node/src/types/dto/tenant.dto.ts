import { InventoryItem, Room, Tenant } from '../';

export interface RoomWithInventory extends Room {
    InventoryItems: InventoryItem[];
}

export interface TenantWithRooms extends Tenant {
    Rooms: RoomWithInventory[];
}
