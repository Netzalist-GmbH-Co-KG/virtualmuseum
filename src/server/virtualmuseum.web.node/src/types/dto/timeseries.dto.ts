import { GeoEvent, GeoEventGroup, TimeSeries } from '../';

export interface GeoEventGroupWithEvents extends GeoEventGroup {
    GeoEvents: GeoEvent[];
}

export interface TimeSeriesWithEvents extends TimeSeries {
    GeoEventGroups: GeoEventGroupWithEvents[];
}
