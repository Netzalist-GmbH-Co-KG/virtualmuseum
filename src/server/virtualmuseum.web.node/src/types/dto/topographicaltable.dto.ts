import {
    TopographicalTable,
    TopographicalTableTopic,
    GeoEvent, 
    GeoEventGroup, 
    TimeSeries
} from '../';

import { MultimediaPresentationWithPresentationItems } from './multimediapresentation.dt';


export interface GeoEventWithMultimediaPresentation extends GeoEvent {
    MultimediaPresentation: MultimediaPresentationWithPresentationItems
}

export interface GeoEventGroupWithEventsAndMultimediaPresentation extends GeoEventGroup {
    GeoEvents: GeoEventWithMultimediaPresentation[],
}

export interface TimeSeriesWithEventsAndMultimediaPresentation extends TimeSeries {
    GeoEventGroups: GeoEventGroupWithEventsAndMultimediaPresentation[],
}

export interface TopographicalTopicWithTimeSeries extends TopographicalTableTopic {
    TimeSeries: TimeSeriesWithEventsAndMultimediaPresentation[],
}

export interface TopographicalTableWithTopics extends TopographicalTable {
    Topics: TopographicalTopicWithTimeSeries[],
}