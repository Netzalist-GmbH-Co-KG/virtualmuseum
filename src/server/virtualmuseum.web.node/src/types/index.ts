export * from './common';
export * from './geo';
export * from './inventory';
export * from './media';
export * from './organization';
