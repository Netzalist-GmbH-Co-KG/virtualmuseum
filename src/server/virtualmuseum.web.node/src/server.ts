import dotenv from 'dotenv';
import { App } from './app';
import './config/swagger.schemas';

// Load environment variables
dotenv.config();

// Create and start the application
const app = new App();
app.start().catch(error => {
    console.error('Failed to start the application:', error);
    process.exit(1);
});
