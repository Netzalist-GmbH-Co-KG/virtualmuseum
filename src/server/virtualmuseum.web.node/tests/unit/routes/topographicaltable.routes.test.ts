import request from 'supertest';
import { App } from '../../../src/app';
import { TestDatabaseService } from '../../helpers/test-database';
import { TopographicalTableWithTopics } from '../../../src/types/dto/topographicaltable.dto';

describe('TopographicalTableRouter', () => {
    let app: App;
    let dbService: TestDatabaseService;

    beforeAll(async () => {
        process.env.NODE_ENV = 'test';
        process.env.API_KEY = 'test-api-key';
        
        // Setup test database
        dbService = new TestDatabaseService();
        await dbService.initialize();

        // Create and initialize app instance with test database
        app = new App(dbService);
        await app.initialize();

        console.log('Test setup complete');
    });

    beforeEach(async () => {
        await dbService.reset();
    });

    afterAll(async () => {
        await app.close();
        console.log('Test cleanup complete');
    });

    describe('GET /api/v1/topographicaltables/:tableId', () => {
        it('should return 401 when API key is missing', async () => {
            const response = await request(app.getApp())
                .get('/api/v1/topographicaltables/test-table-1');
            
            expect(response.status).toBe(401);
            expect(response.body).toEqual({ error: 'Unauthorized' });
        });

        it('should return 401 when API key is invalid', async () => {
            const response = await request(app.getApp())
                .get('/api/v1/topographicaltables/test-table-1')
                .set('x-api-key', 'invalid-key');
            
            expect(response.status).toBe(401);
            expect(response.body).toEqual({ error: 'Unauthorized' });
        });

        it('should return 404 when table does not exist', async () => {
            const response = await request(app.getApp())
                .get('/api/v1/topographicaltables/non-existent-table')
                .set('x-api-key', 'test-api-key');
            
            expect(response.status).toBe(404);
            expect(response.body).toEqual({ error: 'Topographical table not found' });
        });

        it('should return topographical table with topics when API key is valid', async () => {
            const response = await request(app.getApp())
                .get('/api/v1/topographicaltables/test-table-1')
                .set('x-api-key', 'test-api-key');
            
            expect(response.status).toBe(200);
            
            const table: TopographicalTableWithTopics = response.body;
            expect(table).toMatchSnapshot();
        });
    });
});
