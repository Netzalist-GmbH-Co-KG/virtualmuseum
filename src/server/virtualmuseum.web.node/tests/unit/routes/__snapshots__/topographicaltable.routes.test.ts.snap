// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopographicalTableRouter GET /api/v1/topographicaltables/:tableId should return topographical table with topics when API key is valid 1`] = `
{
  "Description": "Test Table Description 1",
  "Id": "test-table-1",
  "Name": "Test Table 1",
  "Topics": [
    {
      "Description": "Test Topic Description 1",
      "Id": "test-topic-1",
      "Name": "Test Topic 1",
      "TimeSeries": [
        {
          "Description": "Test Time Series Description 1",
          "GeoEventGroups": [
            {
              "Description": "Test Group Description 1",
              "GeoEvents": [
                {
                  "DateTime": "2023-01-01T00:00:00Z",
                  "Description": "Test Event Description 1",
                  "GeoEventGroupId": "test-group-1",
                  "Id": "test-event-1",
                  "Label": "Test Event 1",
                  "Latitude": 0,
                  "Longitude": 0,
                  "MultimediaPresentation": {
                    "Description": "Test Presentation Description 1",
                    "Id": "test-presentation-1",
                    "Name": "Test Presentation 1",
                    "PresentationItems": [
                      {
                        "DurationInSeconds": 60,
                        "Id": "test-item-1",
                        "MediaFile": {
                          "Description": "Test Media Description 1",
                          "DurationInSeconds": 60,
                          "FileName": "test1.jpg",
                          "Id": "test-media-1",
                          "Name": "Test Media 1",
                          "Type": 0,
                          "Url": "http://test.com/test1.jpg",
                        },
                        "MediaFileId": "test-media-1",
                        "MultimediaPresentationId": "test-presentation-1",
                        "SequenceNumber": 1,
                        "SlotNumber": 1,
                      },
                      {
                        "DurationInSeconds": 120,
                        "Id": "test-item-2",
                        "MediaFile": {
                          "Description": "Test Media Description 2",
                          "DurationInSeconds": 120,
                          "FileName": "test2.mp4",
                          "Id": "test-media-2",
                          "Name": "Test Media 2",
                          "Type": 1,
                          "Url": "http://test.com/test2.mp4",
                        },
                        "MediaFileId": "test-media-2",
                        "MultimediaPresentationId": "test-presentation-1",
                        "SequenceNumber": 2,
                        "SlotNumber": 2,
                      },
                    ],
                  },
                  "MultimediaPresentationId": "test-presentation-1",
                },
              ],
              "Id": "test-group-1",
              "Label": "Test Group 1",
              "TimeSeriesId": "test-timeseries-1",
            },
          ],
          "Id": "test-timeseries-1",
          "Name": "Test Time Series 1",
        },
      ],
      "TopographicalTableId": "test-table-1",
    },
  ],
}
`;
